[IDE] Overall energy consumption and UI hang improvements

Therzok opened the pull request:
There are various small things added up.

Most notably, the assembly browser is now more responsive and faster to load, and should no longer leak.

Other small improvements

Therzok committed:
Misc perf improvements immediately after project load

Therzok committed:
Use ImmutableDictionary<TargetFramework>, CWT won't work - the strings are collected

Therzok committed:
Dump

Therzok committed:
Use DefaultScheduler via Task.Run, Task.Factory is context dependent

Therzok committed:
[AssemblyBrowser] Cleanups, mostly related to task usage

Therzok committed:
[Debugger, VCS] Some Task.Factory fixes

Fixes VSTS #900427 - Review usage of Task.Factory.StartNew

Therzok committed:
Use ordinal comparisons here

Therzok committed:
[ASsemblyBrowser] Fix child nodes not appearing in the list

Therzok committed:
[VCS] Fix improper usage of cancellation token source

Therzok committed:
Remove unneeded string ordering

Therzok committed:
[AsssemblyBrowser] Fix workspace leak and delay project loading.

It's not common to get to a project item from the IDE, so we can delay loading that while we get everything else setup

Therzok committed:
[MSBuild] Switch this to Task.Run, so we don't accidentally parse on the UI thread

mrward commented on main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/AssemblyBrowserViewContent.cs:
```diff
@@ -169,40 +176,29 @@ public void FindDerivedClasses ()
+
+				var workspace = IdeApp.TypeSystemService.GetWorkspaceAsync (IdeApp.ProjectOperations.CurrentSelectedSolution)
+					.ContinueWith (t => {
+						if (token.IsCancellationRequested)
```
Was wondering if this needed to handle the task exception but it looks like GetWorkspaceAsync would not cause an exception.

Therzok replied:
I had another small patch around this, but it's just a fake hang improvement.

mrward reviewed (approved):
👍

mkrueger commented on main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/AssemblyBrowserWidget.cs:
```diff
@@ -914,24 +914,26 @@ void HPaneExpose (object sender, Gtk.ExposeEventArgs args)
-					} catch (Exception e) {
-						LoggingService.LogError ("Error while opening the assembly browser with id:" + url, e);
+				if (d.IsFaulted) {
+					LoggingService.LogError ("Failed to load assemblies", d.Exception);
```
Shouldn't be a return here ?

Therzok replied:
definitions will be empty if it's faulted, so I guess I can do a return there too.

Therzok replied:
I'm changing a few things here, to better handle faulting.

Therzok replied:
I've thought this through.

There's a few scenarios which are handled by explicitly not returning on faulted.

a) IsFaulted is set when even just one task is faulted, but after all have run. What if we can actually find the type we're looking for? 
b) If the current assembly loader task fails, it'll be caught by the try/catch a bit further below.

Added a comment.

mkrueger commented on main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/TreeNodes/Cecil/TypeDefinitionNodeBuilder.cs:
```diff
@@ -98,33 +98,32 @@ static string GetSource (ITypeDefinition type)
 			bool publicOnly = Widget.PublicApiOnly;
 
-			foreach (var field in type.Fields.OrderBy (m => m.Name, StringComparer.InvariantCulture)) {
+			foreach (var field in type.Fields) {
```
Shouldn't the names be ordered ?

Therzok replied:
Yes, but the sorting is done at treenode level. We don't need to do it here.

Therzok replied:
See the [comparer](https://github.com/mono/monodevelop/blob/5bbe3317491f6f588f5fb7938ed1c1ef8f1169f2/main/src/addins/MonoDevelop.AssemblyBrowser/MonoDevelop.AssemblyBrowser/TreeNodes/AssemblyBrowserTypeNodeBuilder.cs#L60)

Therzok committed:
Add doc comment

Therzok committed:
[AssemblyBrowser] Massively improve time it takes to open and a couple of fixes

Reuse the typesystem created for decompilation. No need to create a new one just to load
the treeview.

Prefer using AddChildren everywhere rather than Add. This helps not sort the tree every time
we're adding an element.

Don't bother gathering all the types in an assembly. Most likely we will only need to gather
types from just one assembly, so defer that to the namespace node builder.

Fix multiple iteration of a type's members when trying to load a project metadata view.

And fix nested types to appear in the tree.

Therzok committed:
Optimize this a bit more

Therzok committed:
Minor refactors to avoid extra Funcs

Therzok committed:
Reuse typesystem for loader, add new feature transform to prevent language breaking

Therzok commented:
@slluis PTAL

Therzok commented:
Rebased and fixed conflicts.
